user/full_user.go

```diff
@@ -0,0 +1,705 @@
+	Hash          string   `json:"-" bson:"userhash,omitempty"`
+	// Private              map[string]*IdHashPair `json:"-" bson:"private"`
+	IsMigrated           bool                `json:"-" bson:"-"`
+	IsUnclaimedCustodial bool                `json:"-" bson:"-"`
+	Enabled              bool                `json:"-" bson:"-"`
+	CreatedTime          string              `json:"createdTime,omitempty" bson:"createdTime,omitempty"`
+	CreatedUserID        string              `json:"createdUserId,omitempty" bson:"createdUserId,omitempty"`
+	ModifiedTime         string              `json:"modifiedTime,omitempty" bson:"modifiedTime,omitempty"`
+	ModifiedUserID       string              `json:"modifiedUserId,omitempty" bson:"modifiedUserId,omitempty"`
+	DeletedTime          string              `json:"deletedTime,omitempty" bson:"deletedTime,omitempty"`
+	DeletedUserID        string              `json:"deletedUserId,omitempty" bson:"deletedUserId,omitempty"`
+	Attributes           map[string][]string `json:"-"`
+	Profile              *UserProfile        `json:"-"`
+	FirstName            string              `json:"firstName,omitempty"`
+	LastName             string              `json:"lastName,omitempty"`
+}
+
+// ExternalUser is the user returned to public services.
+type ExternalUser struct {
+	// same attributes as original shoreline Api.asSerializableUser
+	ID            *string   `json:"userid,omitempty"`
+	Username      *string   `json:"username,omitempty"`
+	Emails        *[]string `json:"emails,omitempty"`
+	EmailVerified *bool     `json:"emailVerified,omitempty"`
+	Roles         *[]string `json:"roles,omitempty"`
+	TermsAccepted *bool     `json:"terms_accepted"`
+}
+
+// MigrationUser is a User that conforms to the structure
+// expected by the keycloak-user-migration keycloak plugin
+type MigrationUser struct {
+	ID            string   `json:"id"`
+	Username      string   `json:"username,omitempty"`
+	Email         string   `json:"email,omitempty"`
+	FirstName     string   `json:"firstName,omitempty"`
+	LastName      string   `json:"lastName,omitempty"`
+	Enabled       bool     `json:"enabled,omitempty"`
+	EmailVerified bool     `json:"emailVerified,omitempty"`
+	Roles         []string `json:"roles,omitempty"`
+	Attributes    struct {
+		TermsAcceptedDate []string `json:"terms_and_conditions,omitempty"`
+	} `json:"attributes"`
+}
+
+/*
+ * Incoming user details used to create or update a `User`
+ */
+type NewUserDetails struct {
+	Username      *string
+	Emails        []string
+	Password      *string
+	Roles         []string
+	EmailVerified bool
+}
+
+type NewCustodialUserDetails struct {
+	Username *string
+	Emails   []string
+}
+
+type UpdateUserDetails struct {
+	Username       *string
+	Emails         []string
+	Password       *string
+	HashedPassword *string
+	Roles          []string
+	TermsAccepted  *string
+	EmailVerified  *bool
+}
+
+type Profile struct {
+	FullName string `json:"fullName"`
+}
+
+var (
```

Comment:
The validation logic doesn't conform to platform validators and should be removed to avoid confusion.